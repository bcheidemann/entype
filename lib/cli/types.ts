import { Lang } from "../emit/types.ts";

export type HelpConfig = {
  help: true;
  invalidArgs: boolean;
  message?: string;
};

export type LangConfig = {
  lang: Lang;
  files: string[];
};

export type Config =
  | HelpConfig
  | LangConfig;
