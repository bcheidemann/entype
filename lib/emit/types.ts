import { Type } from "../types.ts";

export type Lang = "rust" | "typescript";
export type EmitRawFn = (data: string) => void;
export type EmitFn = (
  context: Context,
  name: string,
  type: Type,
  preferInline: boolean,
  code: string,
) => void;
export type EmitAnnotationFn = (annotation: string) => void;
export type GetTypeNameFn = (name?: string) => string;
export type EmitTypeFn = (
  context: Context,
  name: string,
  type: Type,
  preferInline: boolean,
) => string | void;
export type Context = {
  lang: Lang;
  emit: EmitFn;
  emitAnnotation: EmitAnnotationFn;
  emitType: EmitTypeFn;
  getTypeName: GetTypeNameFn;
};
export type EmitterModule = {
  emitType: EmitTypeFn;
};
