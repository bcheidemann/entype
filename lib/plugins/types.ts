import { Context } from "../emit/types.ts";
import { Type } from "../types.ts";

export type PluginName = string;

export type PreEmitHook = (
  context: Context,
  name: string,
  type: Type,
  preferInline: boolean,
  code: string,
) => string | void;

export type Plugin = {
  hooks: {
    preEmitHook?: PreEmitHook;
  };
};
