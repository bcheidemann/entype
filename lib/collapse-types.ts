import {
  isHomogeneousTypeArray,
  isMapType,
  isNullType,
} from "./type-guards.ts";
import {
  ArrayType,
  HomogeneousTypeArray,
  MapType,
  NullType,
  OptionType,
  PrimitiveType,
  StructType,
  Type,
  UnionType,
  UnknownType,
} from "./types.ts";

export function collapseTypes(types: Type[]): Type {
  if (types.length === 0) {
    return { kind: "unknown" };
  }
  return isHomogeneousTypeArray(types)
    ? collapseHomogeneousTypes(types)
    : collapseNonHomogeneousTypes(types);
}

export function collapseHomogeneousTypes(types: HomogeneousTypeArray): Type {
  if (types.length === 0) {
    return { kind: "unknown" };
  }
  switch (types[0].kind) {
    case "struct":
      return collapseStructTypes(types as StructType[]);
    case "union":
      return collapseUnionTypes(types as UnionType[]);
    case "map":
      return collapseMapTypes(types as MapType[]);
    case "option":
      return collapseOptionTypes(types as OptionType[]);
    case "array":
      return collapseArrayTypes(types as ArrayType[]);
    case "primitive":
      return collapsePrimitiveTypes(types as PrimitiveType[]);
    case "null":
      return collapseNullTypes(types as NullType[]);
    case "unknown":
      return collapseUnknownTypes(types as UnknownType[]);
  }
}

export function collapseNonHomogeneousTypes(types: Type[]): Type {
  const uniqueTypes = new Set<Type["kind"]>(
    types.map((type) => type.kind),
  );

  if (uniqueTypes.has("option")) {
    const innerTypes = types
      .map((type) => {
        switch (type.kind) {
          case "option":
            return type.valueType;
          default:
            return type;
        }
      })
      .filter((type) => !isNullType(type));
    const collapsedType = collapseTypes(innerTypes);
    return { kind: "option", valueType: collapsedType };
  }

  if (uniqueTypes.has("union")) {
    return collapseNonHomogeneousTypes(
      types.flatMap((type) => {
        switch (type.kind) {
          case "union":
            return Array.from(type.variants.values());
          default:
            return type;
        }
      }),
    );
  }

  // Coerce nulls to option types
  if (uniqueTypes.has("null")) {
    const nonNullTypes = types.filter((type) => !isNullType(type));
    const collapsedType = collapseTypes(nonNullTypes);
    return { kind: "option", valueType: collapsedType };
  }

  // Coerce structs to maps so that we can collapse them together.
  if (uniqueTypes.has("map") && uniqueTypes.has("struct")) {
    return collapseTypes(
      types.map((type) => {
        switch (type.kind) {
          case "struct":
            return {
              kind: "map",
              valueType: collapseTypes(Array.from(type.fields.values())),
            };
          default:
            return type;
        }
      }),
    );
  }

  // Collapse maps
  if (uniqueTypes.has("map")) {
    const mapTypes = types.filter(isMapType);
    const collapsedMapType = collapseMapTypes(mapTypes);
    const nonMapTypes = types.filter((type) => !isMapType(type));
    return collapseTypes([...nonMapTypes, collapsedMapType]);
  }

  const collectedVariants: {
    struct: StructType[];
    union: UnionType[];
    map: MapType[];
    option: OptionType[];
    array: ArrayType[];
    primitive: PrimitiveType[];
    null: NullType[];
    unknown: UnknownType[];
  } = {
    struct: [],
    union: [],
    map: [],
    option: [],
    array: [],
    primitive: [],
    null: [],
    unknown: [],
  };
  for (const type of types) {
    switch (type.kind) {
      case "struct":
        collectedVariants.struct.push(type);
        break;
      case "union":
        collectedVariants.union.push(type);
        break;
      case "map":
        collectedVariants.map.push(type);
        break;
      case "option":
        collectedVariants.option.push(type);
        break;
      case "array":
        collectedVariants.array.push(type);
        break;
      case "primitive":
        collectedVariants.primitive.push(type);
        break;
      case "null":
        collectedVariants.null.push(type);
        break;
      case "unknown":
        collectedVariants.unknown.push(type);
        break;
    }
  }
  const variants = new Map(
    Object.entries(collectedVariants)
      .filter(([_variant, types]) => types.length > 0)
      .map((
        [variant, types],
      ) => [variant as Type["kind"], collapseHomogeneousTypes(types)]),
  );
  return { kind: "union", variants };
}

export function collapseStructTypes(types: StructType[]): StructType {
  const keys = new Set<string>();
  for (const type of types) {
    for (const key of type.fields.keys()) {
      keys.add(key);
    }
  }
  const fields = new Map<string, Type>();
  for (const key of keys) {
    const fieldType = collapseTypes(
      types.map((type) => type.fields.get(key) ?? { kind: "null" }),
    );
    fields.set(key, fieldType);
  }
  return { kind: "struct", fields };
}

export function collapseUnionTypes(types: UnionType[]): UnionType {
  const variantNames = new Set<Type["kind"] | PrimitiveType["name"]>();
  for (const type of types) {
    for (const variantName of type.variants.keys()) {
      variantNames.add(variantName);
    }
  }
  const variants = new Map<Type["kind"] | PrimitiveType["name"], Type>();
  for (const variantName of variantNames) {
    const variantType = collapseTypes(
      types
        .map((type) => type.variants.get(variantName))
        .filter(Boolean),
    );
    variants.set(variantName, variantType);
  }
  return { kind: "union", variants };
}

export function collapseMapTypes(types: MapType[]): MapType {
  const valueTypes = types.map((type) => type.valueType);
  const valueType = collapseTypes(valueTypes);
  return { kind: "map", valueType };
}

export function collapseOptionTypes(types: OptionType[]): OptionType {
  const valueTypes = types.map((type) => type.valueType);
  const valueType = collapseTypes(valueTypes);
  return { kind: "option", valueType };
}

export function collapseArrayTypes(types: ArrayType[]): ArrayType {
  const elementTypes = types.map((type) => type.elementType);
  const elementType = collapseTypes(elementTypes);
  return { kind: "array", elementType };
}

export function collapsePrimitiveTypes(
  types: PrimitiveType[],
): PrimitiveType | UnionType {
  const uniqueVariants = new Set<PrimitiveType["name"]>();
  for (const type of types) {
    uniqueVariants.add(type.name);
  }
  if (uniqueVariants.size === 1) {
    return { kind: "primitive", name: uniqueVariants.values().next().value };
  }
  const variants = new Map(
    Array
      .from(uniqueVariants.values())
      .map((variant) =>
        [variant, { kind: "primitive", name: variant }] as const
      ),
  );
  return { kind: "union", variants };
}

export function collapseNullTypes(_types: NullType[]): NullType {
  return { kind: "null" };
}

export function collapseUnknownTypes(_types: UnknownType[]): UnknownType {
  return { kind: "unknown" };
}
