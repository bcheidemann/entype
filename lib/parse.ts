import { collapseTypes } from "./collapse-types.ts";
import { Json, JsonArray, JsonObject, Type } from "./types.ts";
import { isIdentifier } from "./util.ts";

export function parseJson(obj: Json): Type {
  switch (typeof obj) {
    case "object":
      if (obj === null) {
        return parseNull(obj);
      }
      if (Array.isArray(obj)) {
        return parseJsonArray(obj);
      }
      return parseJsonObject(obj);
    case "string":
      return { kind: "primitive", name: "string" };
    case "number":
      return { kind: "primitive", name: "number" };
    case "boolean":
      return { kind: "primitive", name: "boolean" };
    default:
      throw new Error("Object is not valid JSON");
  }
}

export function parseNull(_obj: null): Type {
  return { kind: "null" };
}

export function parseJsonArray(obj: JsonArray): Type {
  const elementTypes = obj.map(parseJson);
  const elementType = collapseTypes(elementTypes);
  return { kind: "array", elementType };
}

export function parseJsonObject(obj: JsonObject): Type {
  const fields = new Map(
    Object
      .entries(obj)
      .map(([key, value]) => [key, parseJson(value)])
  );
  const isMap = !Array
    .from(fields.keys())
    .every(isIdentifier);
  if (isMap) {
    const valueTypes = Array.from(fields.values());
    const valueType = collapseTypes(valueTypes);
    return { kind: "map", valueType };
  }
  return { kind: "struct", fields };
}
