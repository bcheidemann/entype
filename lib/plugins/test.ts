import "npm:@total-typescript/ts-reset";
import { describe, it } from "https://deno.land/std@0.198.0/testing/bdd.ts";
import { assertSnapshot } from "https://deno.land/std@0.198.0/testing/snapshot.ts";
import { parseJson } from "../parse.ts";
import { Json, Type } from "../types.ts";
import { collapseTypes } from "../collapse-types.ts";
import { emitRootTypeToString } from "../emit/mod.ts";
import { loadPlugin } from "./mod.ts";

describe("Serde Derive Plugin", () => {
  async function run(
    ctx: Deno.TestContext,
    filename: string,
  ) {
    const json = await Deno.readTextFile(`./fixtures/${filename}`);
    const obj = JSON.parse(json) as Json;
    const type = parseJson(obj);
    const plugins = [await loadPlugin("serde-derive")];
    const code = await emitRootTypeToString("rust", type, plugins);
    await assertSnapshot(ctx, code);
  }

  it("emit array-0.json", async (ctx) => {
    await run(ctx, "array-0.json");
  });

  it("emit array-1.json", async (ctx) => {
    await run(ctx, "array-1.json");
  });

  it("emit array-2.json", async (ctx) => {
    await run(ctx, "array-2.json");
  });

  it("emit array-3.json", async (ctx) => {
    await run(ctx, "array-3.json");
  });

  it("emit array-4.json", async (ctx) => {
    await run(ctx, "array-4.json");
  });

  it("emit boolean-false.json", async (ctx) => {
    await run(ctx, "boolean-false.json");
  });

  it("emit boolean-true.json", async (ctx) => {
    await run(ctx, "boolean-true.json");
  });

  it("emit map-0.json", async (ctx) => {
    await run(ctx, "map-0.json");
  });

  it("emit map-1.json", async (ctx) => {
    await run(ctx, "map-1.json");
  });

  it("emit map-2.json", async (ctx) => {
    await run(ctx, "map-2.json");
  });

  it("emit null.json", async (ctx) => {
    await run(ctx, "null.json");
  });

  it("emit number-negative.json", async (ctx) => {
    await run(ctx, "number-negative.json");
  });

  it("emit number-positive.json", async (ctx) => {
    await run(ctx, "number-positive.json");
  });

  it("emit number-zero.json", async (ctx) => {
    await run(ctx, "number-zero.json");
  });

  it("emit object-0.json", async (ctx) => {
    await run(ctx, "object-0.json");
  });

  it("emit object-1.json", async (ctx) => {
    await run(ctx, "object-1.json");
  });

  it("emit object-2.json", async (ctx) => {
    await run(ctx, "object-2.json");
  });

  it("emit object-3.json", async (ctx) => {
    await run(ctx, "object-3.json");
  });

  it("emit object-4.json", async (ctx) => {
    await run(ctx, "object-4.json");
  });

  it("emit object-5.json", async (ctx) => {
    await run(ctx, "object-5.json");
  });

  it("emit object-6.json", async (ctx) => {
    await run(ctx, "object-6.json");
  });

  it("emit object-7.json", async (ctx) => {
    await run(ctx, "object-7.json");
  });

  it("emit object-8.json", async (ctx) => {
    await run(ctx, "object-8.json");
  });

  it("parses regression-0-option-option.json", async (ctx) => {
    await run(ctx, "regression-0-option-option.json");
  });

  describe("Multiple-Files", () => {
    async function run(
      ctx: Deno.TestContext,
      directory: string,
    ) {
      const files = Deno.readDir(`./fixtures/${directory}`);
      const types = new Array<Type>();
      for await (const file of files) {
        const json = await Deno.readTextFile(
          `./fixtures/${directory}/${file.name}`,
        );
        const obj = JSON.parse(json) as Json;
        types.push(parseJson(obj));
      }
      const type = collapseTypes(types);
      const plugins = [await loadPlugin("serde-derive")];
      const code = await emitRootTypeToString("rust", type, plugins);
      await assertSnapshot(ctx, code);
    }

    it("datapack/blockstates", async (ctx) => {
      await run(ctx, "datapack/blockstates");
    });

    it("datapack/models/block", async (ctx) => {
      await run(ctx, "datapack/models/block");
    });

    it("datapack/models/item", async (ctx) => {
      await run(ctx, "datapack/models/item");
    });
  });
});
