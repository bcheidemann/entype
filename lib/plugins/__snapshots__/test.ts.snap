export const snapshot = {};

snapshot[`Serde Derive Plugin > emit array-0.json 1`] = `
"pub type Root = Vec<f64>;

"
`;

snapshot[`Serde Derive Plugin > emit array-1.json 1`] = `
"#[derive(serde::Serialize, serde::Deserialize)]
pub struct RootElement0 {
  key1: String,
  key2: String,
}

pub type Root = Vec<RootElement0>;

"
`;

snapshot[`Serde Derive Plugin > emit array-2.json 1`] = `
"#[derive(serde::Serialize, serde::Deserialize)]
pub struct RootElement0 {
  key1: String,
  key2: String,
  optional: Option<String>,
}

pub type Root = Vec<RootElement0>;

"
`;

snapshot[`Serde Derive Plugin > emit array-3.json 1`] = `
"pub type Root = Vec<std::collections::HashMap<String, String>>;

"
`;

snapshot[`Serde Derive Plugin > emit array-4.json 1`] = `
"#[derive(serde::Serialize, serde::Deserialize)]
#[serde(untagged)]
pub enum Primitive2 {
  Number(f64),
  String(String),
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct Struct5 {
  key4: String,
}

#[derive(serde::Serialize, serde::Deserialize)]
#[serde(untagged)]
pub enum RootElementEntry1 {
  Primitive(Primitive2),
  Struct(Struct5),
}

pub type Root = Vec<std::collections::HashMap<String, RootElementEntry1>>;

"
`;

snapshot[`Serde Derive Plugin > emit boolean-false.json 1`] = `
"pub type Root = bool;

"
`;

snapshot[`Serde Derive Plugin > emit boolean-true.json 1`] = `
"pub type Root = bool;

"
`;

snapshot[`Serde Derive Plugin > emit map-0.json 1`] = `
"pub type Root = std::collections::HashMap<String, f64>;

"
`;

snapshot[`Serde Derive Plugin > emit map-1.json 1`] = `
"pub type Root = std::collections::HashMap<String, f64>;

"
`;

snapshot[`Serde Derive Plugin > emit map-2.json 1`] = `
"pub type Root = std::collections::HashMap<String, f64>;

"
`;

snapshot[`Serde Derive Plugin > emit null.json 1`] = `
"pub type Root = ();

"
`;

snapshot[`Serde Derive Plugin > emit number-negative.json 1`] = `
"pub type Root = f64;

"
`;

snapshot[`Serde Derive Plugin > emit number-positive.json 1`] = `
"pub type Root = f64;

"
`;

snapshot[`Serde Derive Plugin > emit number-zero.json 1`] = `
"pub type Root = f64;

"
`;

snapshot[`Serde Derive Plugin > emit object-0.json 1`] = `
"#[derive(serde::Serialize, serde::Deserialize)]
pub struct Root {
  key: String,
}

"
`;

snapshot[`Serde Derive Plugin > emit object-1.json 1`] = `
"#[derive(serde::Serialize, serde::Deserialize)]
pub struct Root {
  key: f64,
}

"
`;

snapshot[`Serde Derive Plugin > emit object-2.json 1`] = `
"#[derive(serde::Serialize, serde::Deserialize)]
pub struct Root {
  key: bool,
}

"
`;

snapshot[`Serde Derive Plugin > emit object-3.json 1`] = `
"#[derive(serde::Serialize, serde::Deserialize)]
pub struct Root {
  key: Vec<String>,
}

"
`;

snapshot[`Serde Derive Plugin > emit object-4.json 1`] = `
"#[derive(serde::Serialize, serde::Deserialize)]
pub struct Key0 {
  key: String,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct Root {
  key: Key0,
}

"
`;

snapshot[`Serde Derive Plugin > emit object-5.json 1`] = `
"#[derive(serde::Serialize, serde::Deserialize)]
pub struct Root {
  key: (),
}

"
`;

snapshot[`Serde Derive Plugin > emit object-6.json 1`] = `
"#[derive(serde::Serialize, serde::Deserialize)]
pub struct KeyElement1 {
  key: String,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct Root {
  key: Vec<KeyElement1>,
}

"
`;

snapshot[`Serde Derive Plugin > emit object-7.json 1`] = `
"#[derive(serde::Serialize, serde::Deserialize)]
#[serde(untagged)]
pub enum KeyElement1 {
  Boolean(bool),
  String(String),
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct Root {
  key: Vec<KeyElement1>,
}

"
`;

snapshot[`Serde Derive Plugin > emit object-8.json 1`] = `
"#[derive(serde::Serialize, serde::Deserialize)]
pub struct Key8 {
  key1: String,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct Root {
  key1: bool,
  key2: bool,
  key3: (),
  key4: f64,
  key5: f64,
  key6: String,
  key7: Vec<f64>,
  key8: Key8,
}

"
`;

snapshot[`Serde Derive Plugin > parses regression-0-option-option.json 1`] = `
"#[derive(serde::Serialize, serde::Deserialize)]
pub struct ANDElement3 {
  facing: Option<String>,
  slot_0_occupied: Option<String>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct MultipartElement1 {
  AND: Vec<ANDElement3>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct Root {
  multipart: Vec<MultipartElement1>,
}

"
`;

snapshot[`Serde Derive Plugin > Multiple-Files > datapack/blockstates 1`] = `
"#[derive(serde::Serialize, serde::Deserialize)]
pub struct ArrayElement5 {
  model: String,
  uvlock: Option<bool>,
  weight: Option<f64>,
  x: Option<f64>,
  y: Option<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct Struct15 {
  model: String,
  uvlock: Option<bool>,
  x: Option<f64>,
  y: Option<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
#[serde(untagged)]
pub enum Apply3 {
  Array(Vec<ArrayElement5>),
  Struct(Struct15),
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct TElement29 {
  facing: Option<String>,
  slot_0_occupied: Option<String>,
  slot_1_occupied: Option<String>,
  slot_2_occupied: Option<String>,
  slot_3_occupied: Option<String>,
  slot_4_occupied: Option<String>,
  slot_5_occupied: Option<String>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct TElement66 {
  east: Option<String>,
  north: Option<String>,
  south: Option<String>,
  up: Option<String>,
  west: Option<String>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T24 {
  age: Option<String>,
  AND: Option<Vec<TElement29>>,
  down: Option<String>,
  east: Option<String>,
  facing: Option<String>,
  flower_amount: Option<String>,
  has_bottle_0: Option<String>,
  has_bottle_1: Option<String>,
  has_bottle_2: Option<String>,
  leaves: Option<String>,
  level: Option<String>,
  north: Option<String>,
  OR: Option<Vec<TElement66>>,
  south: Option<String>,
  up: Option<String>,
  west: Option<String>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct TElement2 {
  apply: Apply3,
  when: Option<T24>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct ArrayElement87 {
  model: String,
  x: Option<f64>,
  y: Option<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct Struct93 {
  model: String,
  uvlock: Option<bool>,
  x: Option<f64>,
  y: Option<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
#[serde(untagged)]
pub enum TEntry85 {
  Array(Vec<ArrayElement87>),
  Struct(Struct93),
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct Root {
  multipart: Option<Vec<TElement2>>,
  variants: Option<std::collections::HashMap<String, TEntry85>>,
}

"
`;

snapshot[`Serde Derive Plugin > Multiple-Files > datapack/models/block 1`] = `
"#[derive(serde::Serialize, serde::Deserialize)]
pub struct T5 {
  rotation: Vec<f64>,
  scale: Vec<f64>,
  translation: Vec<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T13 {
  rotation: Vec<f64>,
  scale: Vec<f64>,
  translation: Vec<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T21 {
  rotation: Vec<f64>,
  scale: Vec<f64>,
  translation: Vec<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T29 {
  rotation: Vec<f64>,
  scale: Vec<f64>,
  translation: Vec<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T37 {
  rotation: Vec<f64>,
  scale: Vec<f64>,
  translation: Vec<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T45 {
  rotation: Option<Vec<f64>>,
  scale: Option<Vec<f64>>,
  translation: Vec<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T55 {
  rotation: Vec<f64>,
  scale: Vec<f64>,
  translation: Vec<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T63 {
  rotation: Option<Vec<f64>>,
  scale: Vec<f64>,
  translation: Vec<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T3 {
  firstperson_lefthand: Option<T5>,
  firstperson_righthand: Option<T13>,
  fixed: Option<T21>,
  ground: Option<T29>,
  gui: Option<T37>,
  head: Option<T45>,
  thirdperson_lefthand: Option<T55>,
  thirdperson_righthand: Option<T63>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T78 {
  cullface: Option<String>,
  rotation: Option<f64>,
  texture: String,
  tintindex: Option<f64>,
  uv: Option<Vec<f64>>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T90 {
  cullface: Option<String>,
  rotation: Option<f64>,
  texture: String,
  tintindex: Option<f64>,
  uv: Option<Vec<f64>>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T102 {
  cullface: Option<String>,
  rotation: Option<f64>,
  texture: String,
  tintindex: Option<f64>,
  uv: Option<Vec<f64>>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T114 {
  cullface: Option<String>,
  rotation: Option<f64>,
  texture: String,
  tintindex: Option<f64>,
  uv: Option<Vec<f64>>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T126 {
  cullface: Option<String>,
  rotation: Option<f64>,
  texture: String,
  tintindex: Option<f64>,
  uv: Option<Vec<f64>>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T138 {
  cullface: Option<String>,
  rotation: Option<f64>,
  texture: String,
  tintindex: Option<f64>,
  uv: Option<Vec<f64>>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct Faces76 {
  down: Option<T78>,
  east: Option<T90>,
  north: Option<T102>,
  south: Option<T114>,
  up: Option<T126>,
  west: Option<T138>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T154 {
  angle: f64,
  axis: String,
  origin: Vec<f64>,
  rescale: Option<bool>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct TElement73 {
  __comment: Option<String>,
  faces: Faces76,
  from: Vec<f64>,
  name: Option<String>,
  rotation: Option<T154>,
  shade: Option<bool>,
  to: Vec<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct TElement167 {
  children: Vec<f64>,
  color: f64,
  name: String,
  origin: Vec<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T179 {
  all: Option<String>,
  amethyst: Option<String>,
  back: Option<String>,
  bamboo: Option<String>,
  bar: Option<String>,
  bars: Option<String>,
  base: Option<String>,
  beacon: Option<String>,
  body: Option<String>,
  bottom: Option<String>,
  cactus: Option<String>,
  cactus_top: Option<String>,
  calibrated_side: Option<String>,
  candle: Option<String>,
  cocoa: Option<String>,
  content: Option<String>,
  crop: Option<String>,
  cross: Option<String>,
  dirt: Option<String>,
  down: Option<String>,
  east: Option<String>,
  edge: Option<String>,
  end: Option<String>,
  end_rod: Option<String>,
  eye: Option<String>,
  fan: Option<String>,
  fire: Option<String>,
  flower: Option<String>,
  flowerbed: Option<String>,
  flowerpot: Option<String>,
  front: Option<String>,
  glass: Option<String>,
  glow_lichen: Option<String>,
  hook: Option<String>,
  inner_top: Option<String>,
  inside: Option<String>,
  lantern: Option<String>,
  leaf: Option<String>,
  leg: Option<String>,
  lever: Option<String>,
  line: Option<String>,
  lit: Option<String>,
  lit_log: Option<String>,
  lock: Option<String>,
  log: Option<String>,
  north: Option<String>,
  obsidian: Option<String>,
  overlay: Option<String>,
  pane: Option<String>,
  particle: Option<String>,
  pattern: Option<String>,
  pitcher_bottom: Option<String>,
  pitcher_side: Option<String>,
  pitcher_top: Option<String>,
  pivot: Option<String>,
  plant: Option<String>,
  platform: Option<String>,
  portal: Option<String>,
  post: Option<String>,
  propagule: Option<String>,
  rail: Option<String>,
  round: Option<String>,
  sapling: Option<String>,
  saw: Option<String>,
  sculk_vein: Option<String>,
  side: Option<String>,
  sides: Option<String>,
  slab: Option<String>,
  south: Option<String>,
  stage_1: Option<String>,
  stage_2: Option<String>,
  stage_3_bottom: Option<String>,
  stage_3_top: Option<String>,
  stage_4_bottom: Option<String>,
  stage_4_top: Option<String>,
  stand: Option<String>,
  stem: Option<String>,
  tendrils: Option<String>,
  texture: Option<String>,
  tip: Option<String>,
  top: Option<String>,
  torch: Option<String>,
  tripwire: Option<String>,
  unlit: Option<String>,
  unsticky: Option<String>,
  up: Option<String>,
  upperstem: Option<String>,
  vine: Option<String>,
  wall: Option<String>,
  west: Option<String>,
  wood: Option<String>,
  wool: Option<String>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct Root {
  ambientocclusion: Option<bool>,
  display: Option<T3>,
  elements: Option<Vec<TElement73>>,
  groups: Option<Vec<TElement167>>,
  gui_light: Option<String>,
  parent: Option<String>,
  textures: Option<T179>,
}

"
`;

snapshot[`Serde Derive Plugin > Multiple-Files > datapack/models/item 1`] = `
"#[derive(serde::Serialize, serde::Deserialize)]
pub struct T3 {
  rotation: Vec<f64>,
  scale: Option<Vec<f64>>,
  translation: Vec<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T12 {
  rotation: Vec<f64>,
  scale: Option<Vec<f64>>,
  translation: Vec<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T21 {
  rotation: Option<Vec<f64>>,
  scale: Option<Vec<f64>>,
  translation: Option<Vec<f64>>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T32 {
  rotation: Vec<f64>,
  scale: Option<Vec<f64>>,
  translation: Option<Vec<f64>>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T42 {
  rotation: Option<Vec<f64>>,
  scale: Option<Vec<f64>>,
  translation: Option<Vec<f64>>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T53 {
  rotation: Option<Vec<f64>>,
  scale: Option<Vec<f64>>,
  translation: Vec<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T63 {
  rotation: Vec<f64>,
  scale: Vec<f64>,
  translation: Vec<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T71 {
  rotation: Option<Vec<f64>>,
  scale: Option<Vec<f64>>,
  translation: Vec<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T1 {
  firstperson_lefthand: Option<T3>,
  firstperson_righthand: Option<T12>,
  fixed: Option<T21>,
  ground: Option<T32>,
  gui: Option<T42>,
  head: Option<T53>,
  thirdperson_lefthand: Option<T63>,
  thirdperson_righthand: Option<T71>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T85 {
  texture: String,
  uv: Vec<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct East89 {
  texture: String,
  uv: Vec<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct North93 {
  texture: String,
  uv: Vec<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct South97 {
  texture: String,
  uv: Vec<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct Up101 {
  texture: String,
  uv: Vec<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct West105 {
  texture: String,
  uv: Vec<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct Faces83 {
  down: Option<T85>,
  east: East89,
  north: North93,
  south: South97,
  up: Up101,
  west: West105,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct TElement82 {
  faces: Faces83,
  from: Vec<f64>,
  to: Vec<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct Predicate119 {
  angle: Option<f64>,
  blocking: Option<f64>,
  broken: Option<f64>,
  brushing: Option<f64>,
  cast: Option<f64>,
  charged: Option<f64>,
  filled: Option<f64>,
  firework: Option<f64>,
  level: Option<f64>,
  pull: Option<f64>,
  pulling: Option<f64>,
  throwing: Option<f64>,
  time: Option<f64>,
  tooting: Option<f64>,
  trim_type: Option<f64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct TElement117 {
  model: String,
  predicate: Predicate119,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct T153 {
  layer0: Option<String>,
  layer1: Option<String>,
  layer2: Option<String>,
  particle: Option<String>,
  spyglass: Option<String>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct Root {
  display: Option<T1>,
  elements: Option<Vec<TElement82>>,
  gui_light: Option<String>,
  overrides: Option<Vec<TElement117>>,
  parent: Option<String>,
  textures: Option<T153>,
}

"
`;
